integration-tests/src/cross_chain_transfer.rs

```diff
@@ -460,6 +471,404 @@ fn claim_trapped_asset_should_work() {
+
+		assert_ok!(hydradx_runtime::Tokens::deposit(
+			DOT,
+			&AccountId::from(ALICE),
+			3000 * UNITS
+		));
+
+		assert_ok!(hydradx_runtime::AssetRegistry::set_location(
+			DOT,
+			hydradx_runtime::AssetLocation(MultiLocation::new(1, Here))
+		));
+	});
+
+	Acala::execute_with(|| {
+		let _ = with_transaction(|| {
+			register_dot();
+			TransactionOutcome::Commit(DispatchResult::Ok(()))
+		});
+
+		assert_ok!(hydradx_runtime::Tokens::deposit(
+			DOT,
+			&AccountId::from(ALICE),
+			3000 * UNITS
+		));
+
+		let foreign_asset: MultiAssets = MultiAsset::from((
+			MultiLocation {
+				parents: 1,
+				interior: Junctions::Here,
+			},
+			100 * UNITS,
+		))
+		.into();
+
+		let bob_beneficiary: MultiLocation = Junction::AccountId32 { id: BOB, network: None }.into();
+
+		let xcm =
+			xcm_for_transfer_and_deposit_asset_to_hydra::<hydradx_runtime::RuntimeCall>(foreign_asset, bob_beneficiary);
+
+		//Act
+		let res = hydradx_runtime::PolkadotXcm::execute(
+			hydradx_runtime::RuntimeOrigin::signed(ALICE.into()),
+			Box::new(xcm),
+			Weight::from_parts(399_600_000_000, 0),
+		);
+		assert_ok!(res);
+
+		assert!(matches!(
+			last_hydra_events(2).first(),
+			Some(hydradx_runtime::RuntimeEvent::XcmpQueue(
+				cumulus_pallet_xcmp_queue::Event::XcmpMessageSent { .. }
+			))
+		));
+	});
+
+	//Assert
+	Hydra::execute_with(|| {
+		assert!(matches!(
+			last_hydra_events(2).first(),
+			Some(hydradx_runtime::RuntimeEvent::XcmpQueue(
+				cumulus_pallet_xcmp_queue::Event::Fail { .. }
+			))
+		));
+	});
+}
+
+fn xcm_for_transfer_and_deposit_asset_to_hydra<RC: Decode + GetDispatchInfo>(
+	assets: MultiAssets,
+	beneficiary: MultiLocation,
+) -> VersionedXcm<RC> {
+	use polkadot_runtime::xcm_config::BaseXcmWeight;
+	use xcm_builder::FixedWeightBounds;
+	use xcm_executor::traits::WeightBounds;
+
+	type Weigher<RC> = FixedWeightBounds<BaseXcmWeight, RC, ConstU32<100>>;
+
+	let dest = MultiLocation::new(1, Parachain(HYDRA_PARA_ID));
+
+	let max_assets = assets.len() as u32 + 1;
+	let context = X2(GlobalConsensus(NetworkId::Polkadot), Parachain(ACALA_PARA_ID));
+	let fees = assets
+		.get(0)
+		.expect("should have at least 1 asset")
+		.clone()
+		.reanchored(&dest, context)
+		.expect("should reanchor");
+	let weight_limit = {
+		let fees = fees.clone();
+		let mut remote_message = Xcm(vec![
+			ReserveAssetDeposited::<RC>(assets.clone()),
+			ClearOrigin,
+			BuyExecution {
+				fees,
+				weight_limit: Limited(Weight::zero()),
+			},
+			DepositAsset {
+				assets: Wild(AllCounted(max_assets)),
+				beneficiary,
+			},
+		]);
+		// use local weight for remote message and hope for the best.
+		let remote_weight = Weigher::weight(&mut remote_message).expect("weighing should not fail");
+		Limited(remote_weight)
+	};
+	// executed on remote (on hydra)
+	let xcm = Xcm(vec![
+		//ReserveAssetDeposited(assets.clone()),
+		BuyExecution { fees, weight_limit },
+		DepositAsset {
+			assets: Wild(AllCounted(max_assets)),
```

Comment:
Here is the exact xcm we send. https://github.com/Snowfork/polkadot-sdk/blob/snowbridge/bridges/snowbridge/primitives/router/src/inbound/mod.rs#L259-L271
1. We reserve deposit both DOT and asset being transferred. We always send both DOT and the asset together in one xcm program.
2. Buy execution with DOT only.
3. Deposit the transfer asset only.

Reply:
Thanks for the clarification. So i will send this:

```
	// executed on local (AssetHub)
	let message = Xcm(vec![DepositReserveAsset {
		assets: Definite(vec![dot_as_fee.clone(), assets.clone().into()].into()),
		dest,
		xcm: Xcm(vec![
			// executed on remote (on hydra)
			BuyExecution {
				fees: dot_as_fee,
				weight_limit,
			},
			DepositAsset {
				assets: Definite(assets.into()),
				beneficiary,
			},
		]),
	}]);
```

But some transfer should also happen, right? Because otherwise nothing is in the holding of xcm. That's why I am using `TransferReserveAsset`, but I don't see this in your example. How do you make sure that you have asset in the reserve so you can do DepostiReserveAsset?

Reply:
Thanks for the clarification @alistair-singh .

I am planning to send this message then:
```
	// executed on local (AssetHub)
	let message = Xcm(vec![
		WithdrawAsset(vec![dot_as_fee.clone(), assets.clone().into()].into()),
		BuyExecution {
			fees: dot_as_fee.clone(),
			weight_limit: Unlimited,
		},
		DepositReserveAsset {
			assets: Definite(vec![dot_as_fee.clone(), assets.clone().into()].into()),
			dest,
			xcm: Xcm(vec![
				// executed on remote (on hydra)
				BuyExecution {
					fees: dot_as_fee,
					weight_limit,
				},
				DepositAsset {
					assets: Definite(assets.into()),
					beneficiary,
				},
			]),
		},
	]);
```

Do you think it represents your usecase? At first I really need to do a `WithdrawAsset`, otherwise I have nothing in the holding.

Reply:
This looks correct and supports our case.

`WithdrawAsset` is one way to get assets in holding, but then you would need to create the asset, mint it and send some to a user before you execute the message as that user. Another way you can get assets into holding is by *abusing* the [ReceiveTeleportedAsset](https://github.com/Snowfork/polkadot-sdk/blob/snowbridge/bridges/snowbridge/primitives/router/src/inbound/mod.rs#L247) and [ReserveAssetDeposited](https://github.com/Snowfork/polkadot-sdk/blob/snowbridge/bridges/snowbridge/primitives/router/src/inbound/mod.rs#L251) instructions to force the assets into holding.